File: app/src/main/java/com/yeraygarcia/recipes/adapter/RecipeAdapter.java
package com.yeraygarcia.recipes.adapter;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.constraint.ConstraintLayout;
import android.support.v4.app.FragmentActivity;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.ImageView;
import android.widget.TextView;

import com.yeraygarcia.recipes.R;
import com.yeraygarcia.recipes.RecipeDetailActivity;
import com.yeraygarcia.recipes.RecipeDetailFragment;
import com.yeraygarcia.recipes.database.entity.Recipe;
import com.yeraygarcia.recipes.viewmodel.RecipeViewModel;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Locale;

public class RecipeAdapter extends RecyclerView.Adapter<RecipeAdapter.RecipesViewHolder> implements Filterable {

    private List<Recipe> mRecipes;
    private List<Long> mRecipeIdsInShoppingList = new ArrayList<>();

    private RecipeFilter mFilter;

    private final FragmentActivity mParentActivity;
    private final RecipeViewModel mViewModel;

    // Constructors

    public RecipeAdapter(FragmentActivity parent) {
        mParentActivity = parent;
        mViewModel = ViewModelProviders.of(parent).get(RecipeViewModel.class);
    }

    // Internal classes

    private class RecipeFilter extends Filter {

        private final List<Recipe> originalList;
        private final List<Recipe> filteredList;

        private RecipeFilter(List<Recipe> originalList) {
            super();
            this.originalList = new LinkedList<>(originalList);
            this.filteredList = new ArrayList<>();
        }

        @Override
        protected FilterResults performFiltering(CharSequence charSequence) {
            filteredList.clear();
            final FilterResults results = new FilterResults();

            if (charSequence.length() == 0) {
                filteredList.addAll(originalList);
            } else {
                final String filterPattern = charSequence.toString().toLowerCase().trim();
                for (Recipe recipe : originalList) {
                    if (recipe.getName().toLowerCase().contains(filterPattern)) {
                        filteredList.add(recipe);
                    }
                }
            }

            results.values = filteredList;
            results.count = filteredList.size();
            return results;
        }

        @SuppressWarnings("unchecked")
        @Override
        protected void publishResults(CharSequence charSequence, FilterResults filterResults) {
            setRecipes(filteredList);
        }
    }

    @Override
    public Filter getFilter() {
        if (mFilter == null) {
            mFilter = new RecipeFilter(mRecipes);
        }
        return mFilter;
    }

    class RecipesViewHolder extends RecyclerView.ViewHolder {

        TextView itemName;
        ConstraintLayout itemServingsContainer;
        ImageView itemDecrease;
        TextView itemServings;
        TextView itemServingsLabel;
        ImageView itemIncrease;
        ImageView itemAddToCart;

        private RecipesViewHolder(View itemView) {
            super(itemView);
            itemName = itemView.findViewById(R.id.textview_recipe_name);
            itemServingsContainer = itemView.findViewById(R.id.include_servings);
            itemDecrease = itemView.findViewById(R.id.imageview_decrease_servings);
            itemIncrease = itemView.findViewById(R.id.imageview_increase_servings);
            itemServings = itemView.findViewById(R.id.textview_servings);
            itemServingsLabel = itemView.findViewById(R.id.textview_servings_label);
            itemAddToCart = itemView.findViewById(R.id.imageview_add_to_cart);

            itemName.setOnClickListener(view -> {
                final long recipeId = mRecipes.get(getAdapterPosition()).getId();
                // open Recipe Detail activity
                Intent intent = new Intent(mParentActivity, RecipeDetailActivity.class);
                intent.putExtra(RecipeDetailFragment.ARG_RECIPE_ID, recipeId);
                mParentActivity.startActivity(intent);
            });

            View.OnClickListener addToCartListener = view -> {
                Recipe recipe = mRecipes.get(getAdapterPosition());
                boolean isInShoppingList = mRecipeIdsInShoppingList.contains(recipe.getId());

                if (isInShoppingList) {
                    recipe.increasePortions();
                    mViewModel.updatePortionsInShoppingList(recipe);
                } else {
                    mViewModel.addRecipeToShoppingList(recipe);
                }
            };

            itemAddToCart.setOnClickListener(addToCartListener);
            itemIncrease.setOnClickListener(addToCartListener);
            itemServings.setOnClickListener(addToCartListener);
            itemServingsLabel.setOnClickListener(addToCartListener);

            itemDecrease.setOnClickListener(view -> {
                Recipe recipe = mRecipes.get(getAdapterPosition());
                boolean isInShoppingList = mRecipeIdsInShoppingList.contains(recipe.getId());

                if (isInShoppingList) {
                    if (recipe.getPortions() == 1) {
                        mViewModel.deleteRecipeFromShoppingList(recipe);
                    } else {
                        recipe.decreasePortions();
                        mViewModel.updatePortionsInShoppingList(recipe);
                    }
                }
            });

        }
    }

    @NonNull
    @Override
    public RecipesViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(mParentActivity).inflate(R.layout.item_recipe_name, parent, false);
        return new RecipesViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull RecipesViewHolder holder, int position) {
        if (mRecipes != null) {
            final Recipe currentRecipe = mRecipes.get(position);

            holder.itemName.setText(currentRecipe.getName());
            holder.itemServings.setText(String.format(Locale.getDefault(), "%d", currentRecipe.getPortions()));

            if (currentRecipe.getPortions() == 1) {
                holder.itemDecrease.setImageResource(R.drawable.ic_remove_shopping_cart_24px);
            } else {
                holder.itemDecrease.setImageResource(R.drawable.ic_remove_24px);
            }

            boolean isInShoppingList = mRecipeIdsInShoppingList.contains(currentRecipe.getId());
            holder.itemServingsContainer.setVisibility(isInShoppingList ? View.VISIBLE : View.GONE);
            holder.itemAddToCart.setVisibility(isInShoppingList ? View.GONE : View.VISIBLE);
        } else {
            // Covers the case of data not being ready yet.
            holder.itemName.setText(R.string.no_recipes);
        }
    }

    // getItemCount() is called many times, and when it is first called,
    // mRecipes has not been updated (means initially, it's null, and we can't return null).
    @Override
    public int getItemCount() {
        if (mRecipes != null) {
            return mRecipes.size();
        }
        return 0;
    }

    // Methods

    public List<Recipe> getRecipes() {
        return mRecipes;
    }

    public void setRecipes(List<Recipe> recipes) {
        mRecipes = recipes;
        notifyDataSetChanged();
    }

    public void setRecipeIdsInShoppingList(List<Long> recipeIds) {
        mRecipeIdsInShoppingList = recipeIds;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/com/yeraygarcia/recipes/adapter/RecipeAdapter.kt
package com.yeraygarcia.recipes.adapter

import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.support.constraint.ConstraintLayout
import android.support.v4.app.FragmentActivity
import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.ViewGroup
import android.widget.*
import com.yeraygarcia.recipes.R
import com.yeraygarcia.recipes.RecipeDetailActivity
import com.yeraygarcia.recipes.RecipeDetailFragment
import com.yeraygarcia.recipes.database.entity.Recipe
import com.yeraygarcia.recipes.viewmodel.RecipeViewModel
import kotlinx.android.synthetic.main.item_recipe.view.*
import kotlinx.android.synthetic.main.view_servings_cart.view.*
import java.util.*
import kotlin.collections.ArrayList

class RecipeAdapter(parent: FragmentActivity) :
    RecyclerView.Adapter<RecipeAdapter.RecipesViewHolder>(), Filterable {

    private val context = parent
    private val inflater = parent.layoutInflater
    private val viewModel = ViewModelProviders.of(parent).get(RecipeViewModel::class.java)

    var recipes: List<Recipe> = ArrayList()
        set(value) {
            field = value
            notifyDataSetChanged()
        }
    var recipeIdsInShoppingList: List<UUID> = ArrayList()
        set(value) {
            field = value
            notifyDataSetChanged()
        }
    private val filter: RecipeFilter by lazy { RecipeFilter(recipes) }

    private inner class RecipeFilter(originalList: List<Recipe>) : Filter() {

        private val originalList: List<Recipe> = LinkedList(originalList)
        private val filteredList: MutableList<Recipe> = ArrayList()

        override fun performFiltering(charSequence: CharSequence): Filter.FilterResults {
            filteredList.clear()
            val results = Filter.FilterResults()

            if (charSequence.isEmpty()) {
                filteredList.addAll(originalList)
            } else {
                val filterPattern = charSequence.toString().toLowerCase().trim()
                for (recipe in originalList) {
                    if (recipe.name.toLowerCase().contains(filterPattern)) {
                        filteredList.add(recipe)
                    }
                }
            }

            results.values = filteredList
            results.count = filteredList.size
            return results
        }

        override fun publishResults(
            charSequence: CharSequence,
            filterResults: Filter.FilterResults
        ) {
            recipes = filteredList
        }
    }

    override fun getFilter(): Filter {
        return filter
    }

    inner class RecipesViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        val itemName: TextView = itemView.textview_recipe_name
        val itemDuration: TextView = itemView.textview_recipe_info
        val itemServingsContainer: ConstraintLayout = itemView.include_servings as ConstraintLayout
        val itemDecrease: ImageView = itemView.imageview_decrease_servings
        val itemServings: TextView = itemView.textview_servings
        val itemAddToCart: ImageView = itemView.imageview_add_to_cart
        private val itemServingsLabel: TextView = itemView.textview_servings_label
        private val itemIncrease: ImageView = itemView.imageview_increase_servings
        private val itemNameContainer: LinearLayout = itemView.linearlayout_recipe_name

        init {

            itemNameContainer.setOnClickListener {
                val recipeId = recipes[adapterPosition].id
                // open Recipe Detail activity
                val intent = Intent(context, RecipeDetailActivity::class.java)
                intent.putExtra(RecipeDetailFragment.ARG_RECIPE_ID, recipeId)
                context.startActivity(intent)
            }

            val addToCartListener: (View) -> Unit = {
                val recipe = recipes[adapterPosition]
                val isInShoppingList = recipeIdsInShoppingList.contains(recipe.id)

                if (isInShoppingList) {
                    recipe.increasePortions()
                    viewModel.updatePortionsInShoppingList(recipe)
                } else {
                    viewModel.addRecipeToShoppingList(recipe)
                }
            }

            itemAddToCart.setOnClickListener(addToCartListener)
            itemIncrease.setOnClickListener(addToCartListener)
            itemServings.setOnClickListener(addToCartListener)
            itemServingsLabel.setOnClickListener(addToCartListener)

            itemDecrease.setOnClickListener {
                val recipe = recipes[adapterPosition]

                if (recipeIdsInShoppingList.contains(recipe.id)) {
                    if (recipe.portions == 1) {
                        viewModel.deleteRecipeFromShoppingList(recipe)
                    } else {
                        recipe.decreasePortions()
                        viewModel.updatePortionsInShoppingList(recipe)
                    }
                }
            }

            itemDecrease.setOnLongClickListener {
                val recipe = recipes[adapterPosition]

                if (recipeIdsInShoppingList.contains(recipe.id)) {
                    viewModel.deleteRecipeFromShoppingList(recipe)
                }

                true
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecipesViewHolder {
        val view = inflater.inflate(R.layout.item_recipe, parent, false)
        return RecipesViewHolder(view)
    }

    override fun onBindViewHolder(holder: RecipesViewHolder, position: Int) {
        val currentRecipe = recipes[position]

        holder.itemName.text = currentRecipe.name
        holder.itemDuration.text = currentRecipe.getFormattedDuration(context)
        holder.itemServings.text = String.format(Locale.getDefault(), "%d", currentRecipe.portions)

        if (currentRecipe.portions == 1) {
            holder.itemDecrease.setImageResource(R.drawable.ic_remove_shopping_cart_24px)
        } else {
            holder.itemDecrease.setImageResource(R.drawable.ic_remove_24px)
        }

        val isInShoppingList = recipeIdsInShoppingList.contains(currentRecipe.id)
        holder.itemServingsContainer.visibility = if (isInShoppingList) View.VISIBLE else View.GONE
        holder.itemAddToCart.visibility = if (isInShoppingList) View.GONE else View.VISIBLE
    }

    // getItemCount() is called many times, and when it is first called,
    // mRecipes has not been updated (means initially, it's null, and we can't return null).
    override fun getItemCount(): Int {
        return recipes.size
    }
}
